internal/api/mfa.go

```diff
@@ -388,6 +485,104 @@ func (a *API) challengeTOTPFactor(w http.ResponseWriter, r *http.Request) error
 	})
 }
 
+func (a *API) challengeWebAuthnFactor(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	db := a.db.WithContext(ctx)
+	config := a.config
+
+	user := getUser(ctx)
+	factor := getFactor(ctx)
+	ipAddress := utilities.GetIPAddress(r)
+
+	params := &ChallengeFactorParams{}
+	if err := retrieveRequestParams(r, params); err != nil {
+		return err
+	}
+	if params.WebAuthn == nil {
+		return badRequestError(ErrorCodeValidationFailed, "WebAuthn config required")
+	}
+	webAuthn, err := params.WebAuthn.ToConfig()
+	if err != nil {
+		return err
+	}
+	var response *ChallengeFactorResponse
+	var ws *models.WebAuthnSessionData
+	var challenge *models.Challenge
+	if factor.IsUnverified() {
+		options, session, err := webAuthn.BeginRegistration(user)
+		if err != nil {
+			return internalServerError("error generating WebAuthn registration data").WithInternalError(err)
+		}
+		ws = &models.WebAuthnSessionData{
+			SessionData: session,
+		}
+		challenge = ws.ToChallenge(factor.ID, ipAddress)
+
+		response = &ChallengeFactorResponse{
+			CredentialCreationOptions: options,
+			Type:                      factor.FactorType,
+			ID:                        challenge.ID,
+		}
+
+	} else if factor.IsVerified() {
+		options, session, err := webAuthn.BeginLogin(user)
+		if err != nil {
+			return err
+		}
+		ws = &models.WebAuthnSessionData{
+			SessionData: session,
+		}
+		challenge = ws.ToChallenge(factor.ID, ipAddress)
+		response = &ChallengeFactorResponse{
+			CredentialRequestOptions: options,
+			Type:                     factor.FactorType,
+			ID:                       challenge.ID,
+		}
+
+	}
+
+	err = db.Transaction(func(tx *storage.Connection) error {
+		if terr := factor.WriteChallengeToDatabase(tx, challenge); terr != nil {
+			return terr
+		}
+		return nil
+
+	})
+	if err != nil {
+		return err
+	}
```

Comment:
A single statement doesn't need to be in a transaction.

```suggestion
	if err := factor.WriteChallengeToDatabase(db, challenge); terr != nil {
		return terr
	}
	return nil
```